hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAirdropHandler.java

```diff
@@ -0,0 +1,567 @@
+            newAccountPendingAirdrop = AirdropHandlerHelper.createAccountPendingAirdrop(pendingId, pendingValue);
+        }
+        // Update the sender account with new head pending airdrop
+        var updatedSenderAccount =
+                senderAccount.copyBuilder().headPendingAirdropId(pendingId).build();
+        accountStore.put(updatedSenderAccount);
+        return newAccountPendingAirdrop;
+    }
+
+    /**
+     *  Calculate the fees for the token airdrop transaction. The fees are calculated by combining the
+     *  default airdrop fees, the crypto transfer fees and the token association fees.
+     */
+    @NonNull
+    @Override
+    public Fees calculateFees(@NonNull final FeeContext feeContext) {
+        final var op = feeContext.body().tokenAirdropOrThrow();
+        final var tokensConfig = feeContext.configuration().getConfigData(TokensConfig.class);
+        validateTrue(tokensConfig.airdropsEnabled(), NOT_SUPPORTED);
+
+        final var defaultAirdropFees =
+                feeContext.feeCalculatorFactory().feeCalculator(SubType.DEFAULT).calculate();
+        final var cryptoTransferFees = calculateCryptoTransferFees(feeContext, op.tokenTransfers());
+        final var tokenAssociationFees = calculateTokenAssociationFees(feeContext, op);
+        return combineFees(List.of(defaultAirdropFees, cryptoTransferFees, tokenAssociationFees));
+    }
+
+    private Fees calculateCryptoTransferFees(
+            @NonNull FeeContext feeContext, @NonNull List<TokenTransferList> tokenTransfers) {
+        var cryptoTransferBody = CryptoTransferTransactionBody.newBuilder()
+                .tokenTransfers(tokenTransfers)
+                .build();
+
+        final var syntheticCryptoTransferTxn = TransactionBody.newBuilder()
+                .cryptoTransfer(cryptoTransferBody)
+                .transactionID(feeContext.body().transactionID())
+                .build();
+
+        return feeContext.dispatchComputeFees(syntheticCryptoTransferTxn, feeContext.payer());
+    }
+
+    /**
+     * Calculate the fees for the token associations that need to be created as part of the airdrop. It gathers all the
+     * token associations that need to be created and calculates the fees for each token association.
+     */
+    private Fees calculateTokenAssociationFees(FeeContext feeContext, TokenAirdropTransactionBody op) {
+        // Gather all the token associations that need to be created
+        var tokenAssociationsMap = new HashMap<AccountID, Set<TokenID>>();
+        final var tokenRelStore = feeContext.readableStore(ReadableTokenRelationStore.class);
+        for (var transferList : op.tokenTransfers()) {
+            final var tokenToTransfer = transferList.token();
+            for (var transfer : transferList.transfers()) {
+                if (tokenRelStore.get(transfer.accountID(), tokenToTransfer) == null) {
+                    var list = tokenAssociationsMap.getOrDefault(transfer.accountID(), new HashSet<>());
```

Comment:
This will create a new `HashSet` every call; so `computeIfAbsent()` is cheaper and expresses your intent more clearly than these three lines,
```
tokenAssociationsMap.computeIfAbsent(transfer.accountID(), ignore -> new HashSet<>())
    .add(tokenToTransfer);
```

Reply:
Fixed 👍